### ISSUE TITLE
(rewriter.flat=false) QF_NIA refutation unsoundness

### ISSUE THREAD

**Issue by wintered** [Z3Prover/z3#4837]
Hi,

the following trace shows a refutation unsoundness in Z3 with rewriter.flat=false.
```
[609] % z3release small.smt2
sat
[610] % z3release rewriter.flat=false small.smt2
unsat
[611] % 
[611] % cat small.smt2
(declare-fun a () Int)
(declare-fun b () Int)
(declare-fun c () Int)
(declare-fun d () Int)
(declare-fun e () Int)
(declare-fun f () Int)
(declare-fun g () Int)
(declare-fun h () Int)
(declare-fun i () Int)
(declare-fun j () Int)
(declare-fun k () Int)
(assert
 (and (= (- (* a b (* a d) b c) (* a b c)) (+ a (div (- 1) a)) k 0)
 (>= (+ (* c b) (* e h)) 0)
 (>= (+ (* i a) (* g h f d a) (- 1)) b c j 1 j)))
(check-sat)
[612] %
```
OS: Ubuntu 18.04
Commit: bb24b3f

**Comment by levnach** [comment]
I found a bug it lp_ordered_lemmas and submitted a pull request. But there is more here. When I run "z3 rewriter.flat=false small.smt2 smt.arith.dump_lemmas=true ", which still produces "sat", then the first generated lemma is "sat", according to yices. The lemma is generated from theory_lra::propagate_lp_solver_bound().
here is the lemma
(declare-fun a () Int)
(assert (not (>= (- (mod (- 1) a) a) 0)))
(assert (not (>= (+ (mod (- 1) a) a) 0)))
(assert (>= (mod (- 1) a) 0))
(check-sat)
On this lemma z3 gives "unsat".

**Comment by NikolajBjorner** [comment]
What solution does yices give? If I read it correctly, it encodes the theorem:

```
x < a & x < -a => x < 0
```
So x can only be < a and < -a when x is negative.

**Comment by levnach** [comment]
It gives a = 0

lev@lev-SLaptop:/tmp$ yices --logic=QF_NIA strange.ys 
sat
(= a 0)
lev@lev-SLaptop:/tmp$ cat strange.ys 
(define a :: int)
(assert (not (>= (- (mod (- 1) a) a) 0)))
(assert (not (>= (+ (mod (- 1) a) a) 0)))
(assert (>= (mod (- 1) a) 0))
(check)
(show-model)

**Comment by NikolajBjorner** [comment]
So you fixed the unsoundness bug in Z3, but a lemma we produce shows an unsoundness bug in yices?

**Comment by levnach** [comment]
I fixed a bug it ordered lemmas, but z3 still produces "sat" on "small.smt2". When I was checking the lemmas I got a case where z3 would not respect a timeout. So used yices.  The bug is not fixed yet.

**Comment by NikolajBjorner** [comment]
OK, lemma 106 is SAT, so this should identify the the issue within z3

z3 C:\z3\build\lemma_1444.106.smt2
sat

**Comment by levnach** [comment]
I am on ubuntu seeing lemma...64 as 
(declare-fun a () Int)
(declare-fun c () Int)
(declare-fun b () Int)
(assert (= (div (- 1) a) (- 1)))
(assert (not (>= (+ (* b c) (* (- 1) a b c)) 0)))
(assert (<= (+ a (div (- 1) a)) 0))
(assert (>= (+ a (div (- 1) a)) 0))
(check-sat)

It is unsat, as I can tell: from the first assert a = 1, then, from the second assert, we have bc - bc < 0. Contradiction.
I am running 
z3 rewriter.flat=false small.smt2 smt.arith.dump_lemmas=true

This in the output of git log .
commit 89a6c7a349602c5feb2138627b1194cb587da2cc (HEAD -> master, origin/master, origin/HEAD)
Author: Nikolaj Bjorner <nbjorner@microsoft.com>
Date:   Thu Dec 10 07:29:59 2020 -0800

    fix #4883

**Comment by NikolajBjorner** [comment]
So to conclude, the bug reported here was fixed.
